books/spiders/book_spider.py

```diff
@@ -0,0 +1,76 @@
+import scrapy
+from scrapy.http import Response
+
+BOOK_LIST_CSS = "li.col-xs-6"
+NEXT_PAGE_CSS = "li.next a::attr(href)"
+TITLE_CSS = "div.col-sm-6 h1::text"
+PRICE_CSS = "div.col-sm-6 p.price_color::text"
+AMOUNT_CSS = "div.col-sm-6 p.instock *::text"
+RATING_CSS = "div.col-sm-6 p.star-rating::attr(class)"
+CATEGORY_CSS = "ul.breadcrumb li"
+DESCRIPTION_CSS = "article.product_page p::text"
+UPC_CSS = "table.table tr"
+
+
+class BookSpiderSpider(scrapy.Spider):
+    name = "book_spider"
+    allowed_domains = ["books.toscrape.com"]
+    start_urls = ["https://books.toscrape.com/"]
+
+    def parse(self, response: Response, **kwargs) -> Response:
+        for book in response.css(BOOK_LIST_CSS):
+            detail_page = book.css(
+                "h3 a::attr(href)"
+            ).get()
+            yield response.follow(
+                detail_page, callback=self.parse_book
+            )
+
+        next_page = response.css(
+            NEXT_PAGE_CSS
+        ).get()
+        if next_page is not None:
+            yield response.follow(next_page, callback=self.parse)
+
+    def parse_book(self, response: Response) -> dict:
+        yield {
+            "title": response.css(TITLE_CSS).get(),
+            "price": float(
+                response.css(
+                    PRICE_CSS
+                ).get().replace("£", "")
+            ),
+            "amount_in_stock": int("".join(
+                response.css(AMOUNT_CSS).getall()
+            ).strip().split("(")[1].split()[0]),
+            "rating": self._word_to_number(
+                response.css(
+                    RATING_CSS
+                ).get().split()[1]
+            ),
+            "category": response.css(
+                CATEGORY_CSS
+            ).getall()[2].split('html">')[1].split("</a")[0],
```

Comment:
The way you're extracting the category seems a bit complex and fragile. If the website structure changes, this code could easily break. It would be better to use a more specific CSS selector that directly points to the category element.

Reply:
I can't